### PR TITLE
export classy chalk object to create isolated ctx

### LINKED ISSUE #46: `enabled` is global state
Using chalk from a library I know I always want colours enabled as it is explicitly requested to begin with. But doing `require('chalk').enabled = true` would be bad pratice as this enforce the setting globally and possibly interfere with other libraries using chalk and depending on the auto-detection.

It would be useful to be able to set this option in some isolated way.

### PR DESCRIPTION
**Pull request by keis**
First stab at implementing something like discussed in #46 

wdyt?

-

fixes #46

### CODE REVIEW

**Comment by sindresorhus** [comment]
@keis Thanks for doing this. I've been neglecting making a decision on this for so long.

Would you mind running the benchmark (`npm run bench`) before and after your changes and paste it here?

// @jbnicolai @julien-f @UltCombo @stevenvachon @SBoudrias @naartjie

**Comment by keis** [comment]
Sure thing.

@ master
> chalk@0.5.1 bench /home/keis/git/chalk
> matcha benchmark.js


                      chalk
       1,565,140 op/s » single style
         328,343 op/s » several styles
       1,941,015 op/s » cached styles
         349,328 op/s » nested styles


  Suites:  1
  Benches: 4
  Elapsed: 5,833.32 ms

@ isolated
> chalk@0.5.1 bench /home/keis/git/chalk
> matcha benchmark.js


                      chalk
       1,459,222 op/s » single style
         321,070 op/s » several styles
       2,014,983 op/s » cached styles
         335,849 op/s » nested styles


  Suites:  1
  Benches: 4
  Elapsed: 6,085.44 ms

**Comment by UltCombo** [comment]
Looks nice!
Perhaps export the constructor so one can do `new chalk.Chalk()`? It'd look a bit nicer than accessing the default export's `constructor` property IMHO.

File: index.js [hunk 1]
```diff
@@ -86,14 +93,11 @@ function init() {
...
 }
 
-defineProps(chalk, init());
+defineProps(Chalk.prototype, init());
 
-chalk.styles = ansiStyles;
-chalk.hasColor = hasAnsi;
-chalk.stripColor = stripAnsi;
-chalk.supportsColor = supportsColor;
+Chalk.prototype.styles = ansiStyles;
```
Comment (SBoudrias) [inline index.js]: I don't think any of these needs to be on the prototype, make them Class props.

Comment (keis) [inline index.js]: The idea is to get the names exported so I guess I could just stick them on exported object. Would then change the reference to `this.supportsColor` to just use the local supportsColor as I'm guessing there is not really much use in being able to override that.

**Comment by UltCombo** [comment]
On second thought, seeing as the default export is an instance (which makes sense for back-compat), `new chalk.constructor()` may feel more natural than exposing the constructor as (another) property of the default instance (as I had suggested `new chalk.Chalk()` previously).

Though, the fact that the default export is an instance is pretty much just an implementation detail, in that case `new chalk.Chalk()` would make more sense.

I guess it doesn't matter much either way.

**Comment by sindresorhus** [comment]
@keis This looks good to me. Would you mind adding docs?

Something like:

> The `enabled` option should only be used by the end-user as it affects all chalk consumers. Create a new instance if you need to change this just for your usage.

**Comment by keis** [comment]
@UltCombo I think using the constructor is kinda elegant. But I might be a bit blinded by the novelty of it

@sindresorhus would be good to add this to the chalk.enabled section I guess

>### chalk.enabled
>Color support is automatically detected, but you can override it by setting the
>`enabled` property or by creating a new instance just for your usage. Changing
>the property should only be done from end-user facing applications as it
>affects all chalk consumers.

**Comment by UltCombo** [comment]
I guess how the constructor is exported doesn't matter much.

The text looks a bit confusing though. When you put the "Changing the property should only be done from end-user" as the last sentence it looks like you're talking about all instances. I guess that part should be explicit for the default instance, plugins that create their own instances can do as they please.

**Comment by keis** [comment]
hmm, good point.

"Changing the property should only be done from end-user facing applications as it
affects all consumers of the default chalk instance" maybe?

**Comment by UltCombo** [comment]
Sounds good =]

File: index.js (hunk 1 above)
Comment (sindresorhus) [inline index.js]: :+1:

File: test.js [hunk 2]
```diff
@@ -68,6 +68,14 @@ describe('chalk.enabled', function () {
 	});
 });
 
+describe('chalk.constructor', function () {
+	it('should create a isolated context where colors can be disabled', function () {
+		var ctx = new chalk.constructor(false);
```
Comment (sindresorhus) [inline test.js]: I don't think passing `false` into the constructor is very intuitive. Wouldn't it be better to just create a new instance and then change the `enabled` property?

```js
var ctx = new chalk.constructor();
ctx.enabled = false;
```

Comment (keis) [inline test.js]: Agreed, that looks pretty ugly. But I do like being able to create the instance fully configured in one go.

`new chalk.constructor({enabled: false})` is perhaps cleaner

Comment (sindresorhus) [inline test.js]: :+1:

File: index.js (hunk 1 above)
Comment (sindresorhus) [inline index.js]: @keis Would you mind resolving this so the PR can be landed? :)

Comment (keis) [inline index.js]: yep, I'll get that fixed. thanks poking me was getting was almost forgetting about this :)

**Comment by keis** [comment]
My turn to bump this :)

just added the properties to exports and just now fixed a silly whitespace error

**Comment by jbnicolai** [comment]
Big thumbs up from me :+1:

@sindresorhus any thoughts?

**Comment by sindresorhus** [comment]
Landed! Thanks for pulling through this @keis. We really appreciate it :D

superhighfive

**Comment by keis** [comment]
high five

**Pull request closed without merging**